pkg/csi/cinder/controllerserver.go

```diff
@@ -134,12 +135,28 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 		}
 	}
 
+	// Volume Encryption
+	encrypted := (req.GetParameters()["encrypted"] == "true")
+
+	if encrypted && volType == "" {
+		volType = cinderCSIDefaultEncryptedVolumeType
```

Comment:
I'm not sure which is the right way here. "LUKS" is [documented](https://docs.openstack.org/mitaka/config-reference/block-storage/volume-encryption.html) as the default name but you are right, it's not guaranteed that it exist at all. In that case the API will raise an error which might be fine.

I can add another call to `volumetypes.List()` similar to [`cloud.GetVolumesByName()`](https://github.com/kubernetes/cloud-provider-openstack/blob/master/pkg/csi/cinder/openstack/openstack_volumes.go#L117C22-L117C38) as searching by name is not supported by the `gophercloud` API. This approach would be more expensive than the built in check (as OpenStack uses the [database internally](https://opendev.org/openstack/cinder/src/branch/master/cinder/volume/volume_types.py#L206) to look up by name). As an advantage I can check for the encryption there before creating a volume and error out without the need to cleanup.

Reply:
There's full freedom on the name of the volume types. You shouldn't assume the encrypted one will be named "LUKS". I don't think this PR has the right approach to the problem. What's the difference in just providing the encrypted volume type as the `type` parameter instead?

Instead of relying on a name you could list all the volume types and their encryption information [1] (seems like gophercloud supports this [2]) and make decisions based on that. The problem is - not all clouds will support that and what to do if there are multiple encrypted volume types? Again - I don't understand why relying on setting `type` is not enough to satisfy this use case.

[1] https://docs.openstack.org/api-ref/block-storage/v3/#show-an-encryption-type
[2] https://github.com/gophercloud/gophercloud/blob/ea71744690516fe42b6dba075265587990fdfa09/openstack/blockstorage/v3/volumetypes/results.go#L248-L269